### PR TITLE
Initial version of SpaceService.

### LINKED ISSUE #65: Create an Angular service for querying the Spaces REST API
**Description**

We need an Angular service to interact with the spaces REST API. For now, it would handle querying/listing of spaces.

**Implementation**

This service is imported in the module. It should be available as an injectable service.

**Reference**

User story - #53

### PR DESCRIPTION
**Pull request by VineetReynolds**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

Feature: 

Currently capable of listing all spaces under /api/spaces from the Backend API.
Fixes #65 

* **What is the current behavior?** (You can also link to an open issue here)

No Angular service exists for querying spaces in the backend.

* **What is the new behavior (if this is a feature change)?**

Makes available an injectable service to query spaces.

* **Other information**:

TODO: Additional unit tests for pagination. The `nextLink` property is lost across spaceService invocations. Needs additional work to understand this behavior only in unit-tests.
TODO: Cleaner separation of Space model interface, so prototyping and REST API concerns can be separated. The structure of the existing Space type does not map to the backend API cleanly.

### CODE REVIEW

File: src/app/profile/spaces/space.service.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+  constructor(private http: Http, private logger: Logger, private auth: AuthenticationService) {
+    if (this.auth.getToken() != null) {
+      this.headers.set('Authorization', 'Bearer ' + this.auth.getToken());
+    }
+    logger.log('SpaceService using url:' + this.spacesUrl)
+  }
+
+  getSpaces(pageSize: number = 20): Promise<Space[]> {
+    console.log("Is logger there in getspaces?" + (this.logger == null));
```
Comment (pmuir) [inline src/app/profile/spaces/space.service.ts]: This looks like debugging.

File: src/app/profile/spaces/space.service.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  private spacesUrl = process.env.API_URL + "spaces";
+  private nextLink: string = null;
+
+  private spaces: Space[] = [];
+  private spaceIdIndexMap = {};
+
+  constructor(private http: Http, private logger: Logger, private auth: AuthenticationService) {
+    if (this.auth.getToken() != null) {
+      this.headers.set('Authorization', 'Bearer ' + this.auth.getToken());
+    }
```
Comment (pmuir) [inline src/app/profile/spaces/space.service.ts]: Should this be a reusable function on the AuthenticationService.

File: src/app/profile/spaces/space.service.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    console.log("Is logger there in getspaces?" + (this.logger == null));
+    let url = this.spacesUrl + '?page[limit]=' + pageSize;
+
+    return this.http
+      .get(url, { headers: this.headers })
+      .toPromise()
+      .then(response => {
+        let spaces: Space[];
+        let links = response.json().links;
+        console.log('links:' + links);
```
Comment (pmuir) [inline src/app/profile/spaces/space.service.ts]: More debugging

File: src/app/profile/spaces/space.service.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        if (links.hasOwnProperty('next')) {
+          this.nextLink = links.next;
+          console.log('Setting next link to:' + this.nextLink);
```
Comment (pmuir) [inline src/app/profile/spaces/space.service.ts]: More debugging

File: src/app/profile/spaces/space.service.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    return this.http
+      .get(url, { headers: this.headers })
+      .toPromise()
+      .then(response => {
+        let spaces: Space[];
+        let links = response.json().links;
+        console.log('links:' + links);
+        if (links.hasOwnProperty('next')) {
+          this.nextLink = links.next;
```
Comment (pmuir) [inline src/app/profile/spaces/space.service.ts]: This maybe warrants a code comment?

File: src/app/profile/spaces/space.service.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+  getSpaces(pageSize: number = 20): Promise<Space[]> {
+    console.log("Is logger there in getspaces?" + (this.logger == null));
+    let url = this.spacesUrl + '?page[limit]=' + pageSize;
+
+    return this.http
+      .get(url, { headers: this.headers })
+      .toPromise()
+      .then(response => {
+        let spaces: Space[];
```
Comment (pmuir) [inline src/app/profile/spaces/space.service.ts]: You probably don't need to declare this up here, as you don't use it until you assign lower down.

File: src/app/profile/spaces/space.service.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        let links = response.json().links;
+        console.log('links:' + links);
+        if (links.hasOwnProperty('next')) {
+          this.nextLink = links.next;
+          console.log('Setting next link to:' + this.nextLink);
+        }
+        spaces = response.json().data as Space[];
+        spaces.forEach((item) => {
+          this.logger.log(item);
```
Comment (pmuir) [inline src/app/profile/spaces/space.service.ts]: Probably makes sense to add some context to the log here (what are you logging).

File: src/app/profile/spaces/space.service.ts [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        });
+        // Update the existing workItem big list with new data
```
Comment (pmuir) [inline src/app/profile/spaces/space.service.ts]: Looks like the comment is wrong (should refer to spaces not work items)

File: src/app/profile/spaces/space.service.ts [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        this.updateSpacesList(spaces);
+        return this.spaces;
+      })
+      .catch(this.handleError);
+  }
+
+  getMoreSpaces(): Promise<any> {
```
Comment (pmuir) [inline src/app/profile/spaces/space.service.ts]: Some commonality with the previous method, worth refactoring out?

File: src/app/profile/spaces/space.service.ts [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+          return newSpaces;
+        })
+        .catch(this.handleError);
+    } else {
+      return Promise.reject('No more item found');
+    }
+  }
+
+  private updateSpacesList(spaces: Space[]): void {
+    spaces.forEach((space) => {
```
Comment (pmuir) [inline src/app/profile/spaces/space.service.ts]: A comment here to explain the whole method probably wouldn't go amiss.

File: src/app/profile/spaces/space.service.ts [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+          .splice(this.spaces.length, this.spaces.length, space);
+      }
+    });
+    // Re-build the map once done updating the list
+    this.buildSpaceIndexMapMap();
+  }
+
+  private buildSpaceIndexMapMap(): void {
+    this.spaces.forEach((space, index) => {
+      this.spaceIdIndexMap[space.id] = index;
```
Comment (pmuir) [inline src/app/profile/spaces/space.service.ts]: This might be neater if you use rxjs.

Comment (VineetReynolds) [inline src/app/profile/spaces/space.service.ts]: I'm not quite sure how rxjs would fit in here. Could you please clarify?

**Comment by pmuir** [comment]
I was wondering if it would be neater to do with this with a map operator
or something like that.

File: package.json [hunk 12]
```diff
@@ -102,6 +102,7 @@
     "@angularclass/conventions-loader": "1.0.2",
     "@angularclass/hmr": "1.2.0",
     "@angularclass/hmr-loader": "3.0.2",
+    "@types/lodash": "^4.14.50",
```
Comment (joshuawilson) [inline package.json]: types should go in the devdependencies section

File: src/app/profile/spaces/space.service.spec.ts [hunk 13]
```diff
@@ -0,0 +1,89 @@
+import { cloneDeep } from 'lodash';
```
Comment (joshuawilson) [inline src/app/profile/spaces/space.service.spec.ts]: We are following the same style pattern that the Angular team recommends. Please list the @ angular imports first, then a space followed by the external imports, and another space and internal imports.

File: src/app/profile/spaces/space.service.ts (hunk 2 above)
Comment (joshuawilson) [inline src/app/profile/spaces/space.service.ts]: I'm looking into the Auth service this week. I'll see if it should be added. I can always refactor it out.

**Comment by joshuawilson** [comment]
[test]

File: src/app/profile/spaces/space.service.ts (hunk 2 above)
Comment (VineetReynolds) [inline src/app/profile/spaces/space.service.ts]: I looked at this comment, but put this on the shelf until further discussion. IMHO, the cleanest practice suitable for us, is to use a separate http service that handles bearer token auth, and delegates to the Angular http client service for other actions, like so: http://stackoverflow.com/a/34465070/3916

File: package.json (hunk 12 above)
Comment (VineetReynolds) [inline package.json]: I'm using lodash, for deep cloning, in the SpaceService class, not just in the unit test.

File: src/app/profile/spaces/space.service.spec.ts (hunk 13 above)
Comment (VineetReynolds) [inline src/app/profile/spaces/space.service.spec.ts]: Done. Fixed in both SpaceService and in the unit test.

File: src/app/profile/spaces/space.service.ts (hunk 2 above)
Comment (pmuir) [inline src/app/profile/spaces/space.service.ts]: +1

File: package.json (hunk 12 above)
Comment (joshuawilson) [inline package.json]: Types are for converting TypeScript to JavaScript and are only needed when we compile it, not run it.

**Comment by joshuawilson** [comment]
[test]

**Pull request merged**